cpp/ql/src/semmle/code/cpp/commons/CommonType.qll

```diff
@@ -63,26 +63,287 @@ class Ptrdiff_t extends Type {
   override string getAPrimaryQlClass() { result = "Ptrdiff_t" }
 }
 
+/**
+ * A parent class representing C/C++ a typedef'd `UserType` such as `int8_t`.
+ */
+private abstract class IntegralUnderlyingUserType extends UserType {
+  IntegralUnderlyingUserType() { this.getUnderlyingType() instanceof IntegralType }
+}
+
+/**
+ * A C/C++ fixed-width numeric type, such as `int8_t`.
+ */
+abstract private class TFixedWidthIntegralType extends IntegralUnderlyingUserType { }
+
+class FixedWidthIntegralType extends TFixedWidthIntegralType {
+  FixedWidthIntegralType() { this instanceof TFixedWidthIntegralType }
```

Comment:
This worked!? I'm baffled. Did you know this would work, @MathiasVP? It seems like we can derive from an abstract class without extending it by using `extends T` and `this instanceof T` in combination (without `abstract` it would be redundant).

Reply:
I think this is Pavel's "enshrinkening subtyping" pattern, right? It seems to fit this use-case very well, indeed.

Reply:
Aha. For the record, this pattern was discussed in https://github.slack.com/archives/CP0LHP150/p1606300178023900 (and I was apparently in the discussion).

Reply:
The ever-talented Mr. @lcartey showed me this pattern in a coding session where we were examining all of the options for representing this (and also the history of why we avoid it). We compared it against some other approaches, for example, one based around ADTs, but in the end this one seems to strike a nice balance between convenience and maintainability.